### PR TITLE
TunaNet Integration: MI250x

### PR DESCRIPTION
**Pull request by msaudulhassan**
Enable AI Based Heuristics for Immediate Mode Fallback on MI250x GPUs.

### CODE REVIEW

File: src/conv/heuristics/ai_heuristics.cpp [hunk 1]
```diff
@@ -145,6 +145,47 @@ class Model
     virtual std::vector<float> ToFeatures(const ProblemDescription& problem) const = 0;
 };
 
+class Gfx90aModel : public Model
```
Comment (CAHEK7) [inline src/conv/heuristics/ai_heuristics.cpp]: Just a minor question - should it be `final`?
I guess yes and I guess all of the other models derived from `Model` should be final as well.

File: src/conv/heuristics/ai_heuristics.cpp [hunk 2]
```diff
@@ -145,6 +145,47 @@ class Model
...
+{
+public:
+    Gfx90aModel() : Model("gfx90a") {}
+    bool IsProblemSupported(const ProblemDescription& problem,
+                            const ConvolutionContext& ctx) const override
```
Comment (CAHEK7) [inline src/conv/heuristics/ai_heuristics.cpp]:
```suggestion
    bool IsProblemSupported(const ProblemDescription&,
                            const ConvolutionContext&) const override
```
I guess HIP Tidy complains about unused parameters.

File: src/conv/heuristics/ai_heuristics.cpp [hunk 3]
```diff
@@ -255,7 +296,12 @@ class Gfx908Model : public Model
     }
 };
 
-std::unique_ptr<Model> GetModel(const std::string&) { return std::make_unique<Gfx908Model>(); }
+std::unique_ptr<Model> GetModel(const std::string& device) {
+    if (device == "gfx90a")
+        return std::make_unique<Gfx90aModel>();
+    else
+        return std::make_unique<Gfx908Model>();
```
Comment (CAHEK7) [inline src/conv/heuristics/ai_heuristics.cpp]: What if `device` is not a `gfx90a` and not a `gfx908`? Should it be handled somehow?

File: src/conv/heuristics/ai_heuristics.cpp (hunk 1 above)
Comment (msaudulhassan) [reply src/conv/heuristics/ai_heuristics.cpp]: Good point. I'll mark it final.

File: src/conv/heuristics/ai_heuristics.cpp (hunk 2 above)
Comment (msaudulhassan) [reply src/conv/heuristics/ai_heuristics.cpp]: They are going to be used, I just haven't added the code yet. This PR is in draft.

File: src/conv/heuristics/ai_heuristics.cpp (hunk 3 above)
Comment (msaudulhassan) [reply src/conv/heuristics/ai_heuristics.cpp]: I tried to maintain the previous behavior, but on going through the code again, I think it was a bug. I have fixed it to return a nullpointer if the device is not gfx90a or gfx908

File: src/conv/heuristics/ai_heuristics.cpp (hunk 1 above)
Comment (msaudulhassan) [reply src/conv/heuristics/ai_heuristics.cpp]: @CAHEK7 Done

File: src/conv/heuristics/ai_heuristics.cpp (hunk 3 above)
Comment (msaudulhassan) [reply src/conv/heuristics/ai_heuristics.cpp]: @CAHEK7 Done. I now return a nullptr if the device is not supported

File: src/conv/heuristics/ai_heuristics.cpp [hunk 4]
```diff
@@ -255,7 +255,108 @@ class Gfx908Model : public Model
...
+        {
+            auto solver_id = solver::Id{solver_name.second};
+            auto solver    = solver_id.GetSolver();
+            if(solver.IsApplicable(ctx, problem))
+            {
+                applicable_solvers++;
+                break;
+            }
+        }
+        if(applicable_solvers == 0)
```
Comment (CAHEK7) [inline src/conv/heuristics/ai_heuristics.cpp:310]: Probably it's a bit wordy, but from a high-level point of view if show that we want to find a sovler which `IsApplicable`.

```suggestion
        if(std::find_if(metadata.solver_map.begin(),
                        metadata.solver_map.end(),
                        [&ctx, &problem](const auto& solver_name)
                        {
                            return solver::Id{solver_name.second}.GetSolver().IsApplicable(ctx, problem);
                        }) == metadata.solver_map.end())
```

**Review by CAHEK7 — APPROVED** [review]

**Review by Dmantri98 — APPROVED** [review]

**Review by junliume — APPROVED** [review]
Ping @JehandadKhan for last round of review

File: src/conv/heuristics/ai_heuristics.cpp (hunk 3 above)
Comment (msaudulhassan) [reply src/conv/heuristics/ai_heuristics.cpp]: @CAHEK7 I have revered this back after discussing with JD. If we have a model specifically trained for a given GPU, we use it. If we don't, we use the model for MI100. The assumption is that the fastest solver remains the same no matter the GPU (this is actually not necessarily the case, but we still want to have a heuristic even if we do not have a specially designed model).
